### PR TITLE
Inventory: Restrict access from too far away

### LINKED ISSUE #6658: No protection against long-range inventory access.
1) Start game and open a chest containing a bunch of items.
2) Teleport yourself somewhere far away without closing the chest formspec.
3) Take or put items. You may not necessarily see the inventory changes, but they happen.
4) Works with or without built-in anticheat.

The fix: all inventory accesses should check distance to player (perhaps line-of-sight, too, it it could be done fast enough) before allowing player to modify inventory. Can this be done in mod? Yes. But prefer if done in engine: a) performance, b) everybody gets it without having to do anything.

### PR DESCRIPTION
**Pull request by SmallJoker**
Fixes #6658 by looking up the wielded tool range for each interaction.

### CODE REVIEW

**Review by sfan5 — APPROVED** [review]
tested and works

**Review by rubenwardy — COMMENTED** [review]
The `checkInteractDistance` is kind of messy. Please may you add spacings and reorder statements?

The max_d declaration should be just above the other max_d, and it would be great to comment what the if statement does - "fallback to hand range, or 4 nodes"

**Review by rubenwardy — APPROVED** [review]

**Comment by paramat** [comment]
As long as it doesn't do 'line of sight' concept is ok for me.

**Comment by sfan5** [comment]
There is no "line of sight" check for purposes of anticheat (yet).

**Comment by paramat** [comment]
990d0b0264cae3f05fb0d1353294c94b18faa4c8

**Comment by NathanSalapat** [comment]
Is there anyway for a modder to turn this off for a specific node inventory? I have an inventory I want players to be able to interact with from anywhere in the world. Detached and player inventories won't work for me as I need to let multiple people interact with the same inventory and it needs to persist world restarts.

edit: This might not help at all, as the inventory doesn't update when I remove items, or add items, because the mapblock is too far away. I can still use it, but to a user it looks broken.

**Comment by SmallJoker** [comment]
@NathanSalapat If you do not specify `player_name` for `core.create_detached_inventory`, it can be accessed by everyone. I would recommend you to use that.
EDIT: Use callbacks to periodically store the inventory contents to mod storage or any file in the world directory.

**Comment by appgurueu** [comment]
My two cents: I think what this PR did made sense at the time because people cheating by accessing inventories anywhere was a big problem, but ultimately it's now another (undocumented) game-specific limitation that runs counter to Luanti's goal of being an engine. Tool range is only a crude approximation of whether a player should be allowed access to an inventory.

Perhaps we could get rid of this in the engine in favor of `allow_metadata_inventory_*` callbacks, which could be implemented for chests and furnaces in MTG. The problem of course being that now servers will have come to rely on it.

**Comment by BluebirdGreycoat** [comment]
Since the only way for a player to open a node inventory formspec (outside of mod intervention) is by using a tool/hand to point at it, it continues to make sense that the engine should restrict access to a node inventory that's beyond tool range. To argue otherwise, IMO, is to say that the client should be allowed to script inventory actions without requiring the user to actually open the GUI.

**Comment by appgurueu** [comment]
> Since the only way for a player to open a node inventory formspec (outside of mod intervention) is by using a tool/hand to point at it

This is the case only in Minetest Game and friends. Yes, if you use a node meta formspec with a `list[]` element with an inventory location of `context`, that can only be opened by using a tool with appropriate range, and that should be validated on the server.

But in general, `list[]` allows any inventory location, including `nodemeta:<X>,<Y>,<Z>`. It is possible to `core.show_formspec` make a `list[]` element referencing some node's inventory, and based on our current documentation a modder would expect this to work; there are legitimate use cases for this, say some kind of shared, remote "team chest" every team member should have immediate access to from anywhere.

In other words, *the validation here is simultaneously too strict (in some regards) and too lax (in other regards)*. Proper validation would decide "can the player have a formspec open that references this inventory?", not "does the player wield a tool with sufficient range?". But that's more work, among other things because you need to parse formspecs serverside to some extent. I wrote a thing a couple years ago that did this in a mod: https://github.com/appgurueu/anti_exploit/blob/master/inventory.lua.

Now, I don't think this thing needs to go in the engine, but I think it should at least be possible to conveniently opt out of the current validation reasonably easy, especially if you bring your own, better validation.

**Comment by BluebirdGreycoat** [comment]
I would argue that since nodemeta inventories "look broken" to a user if their mapblock is too far away (as was pointed out above), it doesn't make sense to try to treat a node's inventory--which is inherently spatial--into something globally accessible, i.e., non-spatial. We already have detached inventories for that.

As a modder, a detached inventory would be my first go-to option for an inventory that can be accessed anywhere. 

> Use callbacks to periodically store the inventory contents to mod storage or any file in the world directory.

It's actually easier than that. You can use the on_inventory_action callbacks with a detached inventory to store the detatched inventory's contents into a nodemeta inventory, and have it persist along with the map. No need to touch mod storage or files at all.
